plugins/native/module/vtkF3DSPZReader.cxx

```diff
@@ -0,0 +1,386 @@
+
+      res = inflate(&stream, Z_NO_FLUSH);
+      assert(res == Z_OK || res == Z_STREAM_END);
+
+      uncompressed.insert(
+        uncompressed.end(), buffer.data(), buffer.data() + buffer.size() - stream.avail_out);
+    }
+    inflateEnd(&stream);
+
+    return true;
+  }
+
+  return false;
+}
+
+//----------------------------------------------------------------------------
+struct Header
+{
+  uint32_t magic;
+  uint32_t version;
+  uint32_t numPoints;
+  uint8_t shDegree;
+  uint8_t fractionalBits;
+  uint8_t flags;
+  uint8_t reserved;
+};
+
+//----------------------------------------------------------------------------
+struct PackedCoordinate
+{
+  uint8_t packed[3]; // 24-bits
+
+  float decode(float scale) const
+  {
+    int32_t fixed32 = packed[0];
+    fixed32 |= packed[1] << 8;
+    fixed32 |= packed[2] << 16;
+    fixed32 |= (fixed32 & 0x800000) ? 0xff000000 : 0; // sign
+    return static_cast<float>(fixed32) * scale;
+  }
+};
+
+//----------------------------------------------------------------------------
+struct ColorChannel
+{
+  uint8_t channel;
+
+  uint8_t decode() const
+  {
+    constexpr float colorScale = 0.15f;
+    constexpr float SH_C0 = 0.28209479177387814;
+
+    float fcolor = ((channel / 255.0f) - 0.5f) / colorScale;
+    fcolor = std::clamp(fcolor * SH_C0 + 0.5, 0.0, 1.0);
+
+    return static_cast<uint8_t>(fcolor * 255.0);
+  }
+};
+
+//----------------------------------------------------------------------------
+struct LogScale
+{
+  uint8_t scale;
+
+  float decode() const
+  {
+    return std::exp((scale / 16.0f - 10.0f));
+  }
+};
+
+//----------------------------------------------------------------------------
+struct PackedRotation
+{
+  uint8_t packed[3]; // 24-bits
+
+  std::array<float, 4> decode() const
+  {
+    float xyz[3] = { static_cast<float>(packed[0]) / 127.5f - 1.f,
+      static_cast<float>(packed[1]) / 127.5f - 1.f, static_cast<float>(packed[2]) / 127.5f - 1.f };
+
+    float w = std::sqrt(std::max(0.0f, 1.0f - vtkMath::Dot(xyz, xyz)));
+
+    return { w, xyz[0], xyz[1], xyz[2] };
+  }
+};
+
+//----------------------------------------------------------------------------
+template<int Degree>
+struct SphericalHarmonics
+{
+  uint8_t packed[3 * Degree * (Degree + 2)];
+
+  template<int M, int L>
+  std::array<uint8_t, 3> decode() const
+  {
+    static_assert(L > 0 && L <= Degree);
+    static_assert(M <= L && M >= -L);
+
+    constexpr int offset = 3 * ((M + L) + L * L - 1);
+
+    std::array<uint8_t, 3> sh;
+    for (int i = 0; i < 3; i++)
+    {
+      sh[i] = packed[offset + i];
+    }
+    return sh;
+  }
+};
+
+//----------------------------------------------------------------------------
+template<int Degree>
+void AddSphericalHarmonics(int nbSplats, unsigned char* buffer, vtkPointData* pointData)
+{
+  SphericalHarmonics<Degree>* begin =
+    reinterpret_cast<SphericalHarmonics<Degree>*>(buffer + 16 + (9 + 4 + 3 + 3) * nbSplats);
+
+  auto getSuffix = [](int m) -> std::string
```

Comment:
what does m stand for here ?

Reply:
`M` and `L` are common naming for spherical harmomics. See https://en.wikipedia.org/wiki/Table_of_spherical_harmonics#/media/File:Real_Spherical_Harmonics_Figure_Table_Complex_Polar_Plot.gif

Reply:
fair.